docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,214 @@
+    /// arguments instead of relying on the CLR to generate an IL Stub at runtime.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public sealed class GeneratedDllImportAttribute : Attribute
+    {
+        /// <summary>
+        /// Enables or disables best-fit mapping behavior when converting Unicode characters
+        /// to ANSI characters.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.BestFitMapping"/>
+        public bool BestFitMapping;
+
+        /// <summary>
+        /// Indicates the calling convention of an entry point.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.CallingConvention"/>
+        public CallingConvention CallingConvention;
+
+        /// <summary>
+        /// Indicates how to marshal string parameters to the method and controls name mangling.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.CharSet"/>
+        public CharSet CharSet;
+
+        /// <summary>
+        /// Indicates the name or ordinal of the DLL entry point to be called.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.EntryPoint"/>
+        public string? EntryPoint;
+
+        /// <summary>
+        /// Controls whether the System.Runtime.InteropServices.DllImportAttribute.CharSet
+        /// field causes the common language runtime to search an unmanaged DLL for entry-point
+        /// names other than the one specified.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.ExactSpelling"/>
+        public bool ExactSpelling;
+
+        /// <summary>
+        /// Indicates whether unmanaged methods that have HRESULT or retval return values
+        /// are directly translated or whether HRESULT or retval return values are automatically
+        /// converted to exceptions.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.PreserveSig"/>
+        public bool PreserveSig;
+
+        /// <summary>
+        /// Indicates whether the callee calls the SetLastError Windows API function before
+        /// returning from the attributed method.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.SetLastError"/>
+        public bool SetLastError;
+
+        /// <summary>
+        /// Enables or disables the throwing of an exception on an unmappable Unicode character
+        /// that is converted to an ANSI "?" character.
+        /// </summary>
+        /// <see cref="System.Runtime.InteropServices.DllImportAttribute.ThrowOnUnmappableChar"/>
+        public bool ThrowOnUnmappableChar;
+    }
+}
+```
+
+## Questions
+
+* Can the above API be used to provide a reverse P/Invoke stub?
```

Comment:
Today, reverse P/Invoke stubs are done via delegates. Inserting the marshaling stub for the delegate would not be easy (in most cases at least).

A good starting position may be that delegates are not supported with this, and need to switch to function pointers and NativeCallabable methods instead.